fem/libceed/solvers-atpmg.cpp

```diff
@@ -0,0 +1,728 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "solvers-atpmg.h"
+#include "util.h"
+
+#ifdef MFEM_USE_CEED
+#include <ceed-backend.h>
+
+#include <math.h>
+// todo: should probably use Ceed memory wrappers instead of calloc/free?
+#include <stdlib.h>
+
+int coarse_1d_edof(int i, int P1d, int coarse_P1d)
+{
+   int coarse_i = (i < coarse_P1d - 1) ? i : -1;
+   if (i == P1d - 1)
+   {
+      coarse_i = coarse_P1d - 1;
+   }
+   return coarse_i;
+}
+
+int reverse_coarse_1d_edof(int i, int P1d, int coarse_P1d)
+{
+   int coarse_i;
+   if (i > P1d - coarse_P1d)
+   {
+      coarse_i = i - (P1d - coarse_P1d);
+   }
+   else
+   {
+      coarse_i = -1;
+   }
+   if (i == 0)
+   {
+      coarse_i = 0;
+   }
+   return coarse_i;
+}
+
+int min4(int a, int b, int c, int d)
+{
+   if (a <= b && a <= c && a <= d)
+   {
+      return a;
+   }
+   else if (b <= a && b <= c && b <= d)
+   {
+      return b;
+   }
+   else if (c <= a && c <= b && c <= d)
+   {
+      return c;
+   }
+   else
+   {
+      return d;
+   }
+}
+
+int CeedATPMGElemRestriction(int order,
+                             int order_reduction,
+                             CeedElemRestriction er_in,
+                             CeedElemRestriction* er_out,
+                             CeedInt *&dof_map)
+{
+   int ierr;
+   Ceed ceed;
+   ierr = CeedElemRestrictionGetCeed(er_in, &ceed); CeedChk(ierr);
+
+   CeedInt numelem, numnodes, numcomp, elemsize;
+   ierr = CeedElemRestrictionGetNumElements(er_in, &numelem); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetLVectorSize(er_in, &numnodes); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetElementSize(er_in, &elemsize); CeedChk(ierr);
+   ierr = CeedElemRestrictionGetNumComponents(er_in, &numcomp); CeedChk(ierr);
+   if (numcomp != 1)
+   {
+      // todo: this will require more thought
+      return CeedError(ceed, 1, "Not implemented!");
+   }
+
+   int P1d = order + 1;
+   int coarse_P1d = P1d - order_reduction;
+   int dim = (log((double) elemsize) / log((double) P1d)) + 1.e-3;
+
+   CeedVector in_lvec, in_evec;
+   ierr = CeedElemRestrictionCreateVector(er_in, &in_lvec, &in_evec);
+   CeedChk(ierr);
+
+   CeedScalar * lvec_data;
+   ierr = CeedVectorGetArray(in_lvec, CEED_MEM_HOST, &lvec_data); CeedChk(ierr);
+   for (int i = 0; i < numnodes; ++i)
+   {
+      lvec_data[i] = (CeedScalar) i;
+   }
+   ierr = CeedVectorRestoreArray(in_lvec, &lvec_data); CeedChk(ierr);
+
+   // todo: I am making assumptions about the ordering of the evec that in
+   // principle are decided by the backend, which I do not control
```

Comment:
Could this result in silent bugs?

Reply:
Yes, probably. It seems to work for all the backends I have tested.

Reply:
Evec ordering possibly being backend dependant is a core assumption in libCEED's internals

Reply:
Is it entirely arbitrary at the global evector level? Can I at least assume that each element's dofs are placed together in the evector?

Reply:
https://libceed.readthedocs.io/en/latest/libCEEDdev/#internal-layouts

Reply:
Let me know if I can clarify anything in that - it's buried deep in our docs and not a lot of people who aren't me have read it

Reply:
That's already very helpful, and will allow me to (a) check for the correct layout in some cases and (b) at least document what we're expecting so it's clear what's going on.

Reply:
Clarification/Question: `CeedElemRestrictionCreate` takes an `offsets` array and documents its layout, which is independent of the backend. In practice I have found this layout is usually the same as the evector layout, but I know from the above this is not guaranteed. I guess I want to verify that `CeedElemRestrictionCreate` always takes its input in the same layout regardless of backend (ie, the backend might reorder it) and that the backend should call `CeedElemRestrictionSetELayout` to tell me how/if it reordered things.

Reply:
Yes, the `offsets` array always has the same structure.
```
  CeedInt layout[3] = {1, elemsize*nelem, elemsize};
  ierr = CeedElemRestrictionSetELayout(r, layout); CeedChk(ierr);
```
The vanilla CUDA backend uses the above layout, while the all CPU backends use
```
  CeedInt layout[3] = {1, elemsize, elemsize*ncomp};
  ierr = CeedElemRestrictionSetELayout(r, layout); CeedChk(ierr);
```
In theory, a backend need not use an evector that can be described like this. We don't have any yet, but @YohannDudouit has discussed creating a backend that does not use an evector layout that can be described in this way.

Reply:
If some future backend does not use this organization, does `CeedElemRestrictionGetELayout()` return a nonzero error code? If so I can handle this situation relatively gracefully (it won't work, but at least the user knows it won't work).

Reply:
We haven't decided how that would work yet. We are about to revamp the error system, so the idea of recoverable errors might be added.

Currently though, you'd just get the array `[0, 0, 0]` back for such a backed.

Reply:
woops, I misrememberd

https://github.com/CEED/libCEED/blob/954a603368e46bb36631709b22e7d1287756ce4e/interface/ceed-elemrestriction.c#L202

We actually return a non-zero error code when there has been no ELayout set, but depending upon the error handler in use, this can be an abort.